File: houseRobberRecursive.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Time Limit Exceeded
class Solution {
    public int rob(int[] nums) {
        return helper(nums, 0, 0);
    }
    private int helper(int[] nums, int index, int amount){
        if(index>=nums.length) return amount;
        int case1 = helper(nums, index+1, amount); //house is not robbed
        int case2 = helper(nums, index+2, nums[index] + amount); //house is robbed
        return Math.max(case1,case2);
    }
}

File: HouseRobberRecursive.py
# // Time Complexity : 2^N
# // Space Complexity : 2^N
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this :


# // Your code here along with comments explaining your approach
class Solution:
    def rob(self, nums: List[int]) -> int:
        if nums == None:
            return 0
        return self.helper(nums,0,0)

    def helper(self,nums, index, amount):
        #base
        if index >= len(nums):
            return amount
        
        
        # choose the house
        case1 = self.helper(nums, index + 2, amount + nums[index])
        
        # not choose the house
        case2 = self.helper(nums, index + 1, amount)
        
        return max(case1, case2)
    
        
        